src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,910 @@
+				if err := m.TryClearFieldByNumber(fieldNumInt); err != nil {
+					return fmt.Errorf("error: %v clearing field: %d", err, fieldNumInt)
+				}
+			} else {
+				isDefaultValue, err := isDefaultValue(field, curVal)
+				if err != nil {
+					return fmt.Errorf(
+						"error: %v, checking if %v is default value for field %s",
+						err, curVal, field.String())
+				}
+				if isDefaultValue {
+					fieldsChangedToDefault = append(fieldsChangedToDefault, fieldNum)
+				}
+
+				changedFields = append(changedFields, fieldNum)
+				if err := enc.lastEncoded.TrySetFieldByNumber(fieldNumInt, curVal); err != nil {
+					return fmt.Errorf(
+						"error: %v setting field %d with value %v on lastEncoded",
+						err, fieldNumInt, curVal)
+				}
+			}
+		}
+	}
+
+	if len(changedFields) == 0 && enc.lastEncoded != nil {
+		// Only want to skip encoding if nothing has changed AND we've already
+		// encoded the first message.
+		enc.stream.WriteBit(opCodeNoChange)
+		return nil
+	}
+
+	// TODO(rartoul): Need to add a MarshalInto to the ProtoReflect library to save
+	// allocations: https://github.com/m3db/m3/issues/1471
+	marshaled, err := m.Marshal()
+	if err != nil {
+		return fmt.Errorf("%s error trying to marshal protobuf: %v", encErrPrefix, err)
+	}
+
+	// Control bit indicating that proto values have changed.
+	enc.stream.WriteBit(opCodeChange)
+	if len(fieldsChangedToDefault) > 0 {
+		// Control bit indicating that some fields have been set to default values
+		// and that a bitset will follow specifying which fields have changed.
+		enc.stream.WriteBit(opCodeFieldsSetToDefaultProtoMarshal)
+		enc.encodeBitset(fieldsChangedToDefault)
+	} else {
+		// Control bit indicating that none of the changed fields have been set to
+		// their default values so we can do a clean merge on read.
+		enc.stream.WriteBit(opCodeNoFieldsSetToDefaultProtoMarshal)
+	}
+	enc.encodeVarInt(uint64(len(marshaled)))
+	enc.stream.WriteBytes(marshaled)
+
+	if enc.lastEncoded == nil {
+		// Set lastEncoded to m so that subsequent encodings only need to encode fields
+		// that have changed.
+		enc.lastEncoded = m
+	} else {
+		// lastEncoded has already been mutated to reflect the current state.
+	}
+
+	return nil
+}
+
+func (enc *Encoder) encodeFirstSignedIntValue(i int, v int64) {
+	neg := false
+	enc.customFields[i].prevIntBits = uint64(v)
+	if v < 0 {
+		neg = true
+		v = -1 * v
+	}
+
+	vBits := uint64(v)
+	numSig := encoding.NumSig(vBits)
+
+	enc.customFields[i].intSigBitsTracker.WriteIntSig(enc.stream, numSig)
+	enc.encodeIntValDiff(vBits, neg, numSig)
+}
+
+func (enc *Encoder) encodeFirstUnsignedIntValue(i int, v uint64) {
+	enc.customFields[i].prevIntBits = v
+
+	numSig := encoding.NumSig(v)
+	enc.customFields[i].intSigBitsTracker.WriteIntSig(enc.stream, numSig)
+	enc.encodeIntValDiff(v, false, numSig)
+}
+
+func (enc *Encoder) encodeNextSignedIntValue(i int, next int64) {
+	prev := int64(enc.customFields[i].prevIntBits)
+	diff := next - prev
```

Comment:
Is it worth extracting this out eventually so we can share an int encoder with m3tsz? Will the diff ordering being reversed make that impossible in the future with existing data? It's def more logical, but you may want to do the same direction as m3tsz just so it remains possible to create a common in encoder in the future (unless there are other reasons why that won't be possible).